src/execute/NumericExecutionPass.cpp

```diff
@@ -2078,6 +2113,13 @@ u1 NumericExecutionPass::run( libpass::PassResult& pr )
         InvariantChecker invariantChecker( locationRegistry, globalState );
         invariantChecker.check( *specification );
 
+        if( not scheduler.check() )
```

Comment:
Move this warning after the while loop, and use `scheduler.numberOfSteps() == 0` condition instead. The `check` method can then be dropped.

Also another warning message would be good, because the current one is not 100% correct, e.g. `init {(0)-> undef}`. -> there is an init definition, but no agent is initially available.

Maybe: "Could not perform a single step because no agent was initially available. This may happen when no valid initial rule has been specified (see InitRule)."

Reply:
good point, could you please change this so that we can finalize this PR? thx

Reply:
The `scheduler.numberOfSteps() == 0` is correct in this cause, because when no agents are available the scheduler doesn't make a step, see https://github.com/casm-lang/libcasm-fe/blob/master/src/execute/NumericExecutionPass.cpp#L1888

Reply:
true, but the problem and therefore the implementation of the `check` was that the `collectAgents` method results into an error for not picking up a `IR::ReferenceConstant` because the "no init" results into a `VoidConstant`, just try it out, you will see :smile:

Reply:
Which means that `check` has hidden the real problem. If you get a void constant where a reference is expected means that type preservation is violated.

Reply:
There went something wrong during rebase. Dropped commits have been re-added and these changes (+ unrelated changes in FrameSizeDeterminationVisitor) are included again.